### ISSUE TITLE
🧙‍♂️ Adaptive Caching

### ISSUE THREAD

**Issue by jodydonetti** [jodydonetti/ZiggyCreatures.FusionCache#24]
## Scenario

Almost every FusionCache method that act on cache entries - like `Set`, `Remove`, `GetOrSet`, etc - can receive various options via a param of type `FusionCacheEntryOptions` which contains things like `Duration`, `Priority`, `FailSafeMaxDuration` and so on.

This is all well and good but in the case of the `GetOrSet` method, there could be one more opportunity: the abilty to **adapt** the caching options to the value about to be cached when returned by the factory, if and when that is called.

## Examples

For example if our factory retrieves a `Product` from a database, the cache duration may be small in case the product has just been created/updated (since it may receive additional updates soon) or very large in case a product is old and no more available to sell (since it most probably won't change anymore).
In this case the information to base the adaptive logic would be inside the product instance itself, making it self-contained.

A different example can be a resource, let's say a `News` piece in an online newspaper, gathered not via a database call but via a remote webservice call: in the http response we may have a `Cache-Control` header which can drive the cache duration inside FusionCache itself.
In this case the information to base the adaptive logic would NOT be inside the news instance itself, making it not self-contained.

## Solution

I want to add a way to change the `FusionCacheEntryOptions` param passed to a `GetOrSet` method call based on the value produced by a factory (or adjacent data, see the previous News example).

There are different ways to achieve this, and I'd like to highlight them, with pros and cons.

|**💡 NOTE**|
|:-------------|
|Regardless of which proposal will be choosen, new overloads will be added to match the currently available methods, so that method calls in already existing code will not need to be fixed.|

## Proposal 01: additional lambda

The idea here is to have an additional param of type `Action<TValue, FusionCacheEntryOptions>` so to be able to have a way to receive the value produced by the factory + the options param currently being used and change something there.

An example can be something like this:

```csharp
var product = cache.GetOrSet<Product>(
    $"product:{id}",
    _ => GetProductFromDb(id),
    options => options.SetDuration(TimeSpan.FromMinutes(1)),
    (product, options) => {
      // IF THE PRODUCT IS EXPIRED -> CHANGE THE DURATION TO 1 HOUR
      if (product.ExpirationDate < DateTimeOffset.UtcNow) {
        options.SetDuration(TimeSpan.FromHours(1));
      }
    }
);
```

**👍 PROS**:
- we can easily take existing code and just add a new param to existing method calls to make the options adaptive
- since we are talking about a separate param, we can define one or more and reuse the same delegate instance multiple times (eg: `public static ProductUtils.AdaptCachingOptions(Product product, FusionCacheEntryOptions options)` )
- since the adaptive part here would be explicit, we would be easily able to save some allocations and avoid unwanted options change (see the "⚠ Warning" section below)

**👎 CONS**:
- an additional lambda would need to be allocated, but only in case adaptivity is needed: when not needed, there would be no new allocations
- in case the piece of data needed to adapt the options is not part of the returned value itself (eg: a Product instance) there would be no way to pass that to the adaptive logic. Practical example: the remote source for the factory is a webservice, and we would like to use the `Cache-Control` header of the http response to drive the cache duration: in the factory we would have access to the http response, before deserializing the response body to a Product instance, but not anymore in the separate lambda

**ℹ NOTES**:
- the additional param would be optional with a default value of `null` which would indicate not to change the options, basically disabling _adaptivity_ (and the extra allocation)

## Proposal 02: new factory signature

The idea is to change the common signature for the factory param, from:

```csharp
Func<CancellationToken, Task<TValue>>
```

to:

```csharp
Func<CancellationToken, FusionCacheEntryOptions, Task<TValue>>
```

Again, new overloads to map the currently existing factory signature would be created, so existing code would continue working as usual.

**👍 PROS**:
- having everything directly in the same lambda (the factory itself) would make it possible to adapt the options based on data not part of the return value (eg: an http header not part of a Product instance, like in the 2nd cons of Proposal 01)
- the signature would be quite similar to what is currently used in the common MemoryCache interface, therefore possibly lowering the entry barrier for new users
- in theory the extra lambda allocation in the proposal 01 would not be needed, but...

**👎 CONS**:
- ... but, since internally the implementation would be just one (with the factory param with the new signature), any call with the old signature would need to be automatically "wrapped" in a lambda created on the fly that would turn one call into the other (eg: `(token, options) => factory(token)` ) and there it is another invisible allocation
- since the adaptive part here would NOT be explicit, we wouldn't know if a factory will change options or not, making it harder to save some allocations and avoid unwanted options change (see the "⚠ Warning" section below)

**ℹ NOTES**:
- A possible way out of this would be to mark all the new overloads that map the old factory definition to the new one as `[Obsolete]` with an explanation message to use the new signature, and just move on. But I'm not that sure this is the right way to handle this situation, I'm still thinking about this

## Proposal 03 (additional): new `IObjectWithFusionCacheAdaptiveOptions` interface

This is not a substitute for any of the 2 previous proposals, but may be a nice addition.
The idea is to have a new `IObjectWithFusionCacheAdaptiveOptions` interface modeled like this:

```csharp
public interface IObjectWithFusionCacheAdaptiveOptions
{
  void AdaptFusionCacheEntryOptions(FusionCacheEntryOptions options);
}
```

and it would be used, following the Product example above, like this:

```csharp
public class Product
  : IObjectWithFusionCacheAdaptiveOptions
{

  [...]

  public DateTimeOffset ExpirationDate { get; set; }

  [...]

  public void AdaptFusionCacheEntryOptions(FusionCacheEntryOptions options)
  {
    // IF THE PRODUCT IS EXPIRED -> CHANGE THE DURATION TO 1 HOUR
    if (ExpirationDate < DateTimeOffset.UtcNow) {
      options.SetDuration(TimeSpan.FromHours(1));
    }
  }
}
```

**👍 PROS**:
- quick and easy to use
- in case proposal 1 is choosen, this can be used as a fallback when no logic is passed explicitly via the extra param. Otherwise that would win (since it would be explicitly passed per-call, instead of being per-type). As a comparison, think of this as the before/after serialization callbacks approach in Json.NET.

**👎 CONS**:
- none actually, since it's an **extra** way to achieve the goal

**ℹ NOTES**:
- it may be considered a little bit more invasive, since you would have to directly touch your own classes. But, and this is important, since it would be an **additional** way to get the result, I don't see it as a problem, really
- if this would be the only way to do this, the problem would be that it would not be usable with types we don't control, with sealed types or with primitive types. But since this would be just an **additional** way to do it, I see no problem with this approach also being available

## ⚠ Warning

Special care must be put into **NOT** changing a `FusionCacheEntryOptions` instance when that is not the intention.

What do I mean?

A good way to save some allocations is to define and reuse the same options object over and over again, passing it to the same method call all the time.

Example:

```csharp
public static class CachingDefaults
{
  public static FusionCacheEntryOptions ProductOptions = new FusionCacheEntryOptions()
  {
    Duration = TimeSpan.FromMinutes(1),
    IsFailSafeEnabled = true,
    FailSafeMaxDuration = TimeSpan.FromHours(24),
    FailSafeThrottleDuration = TimeSpan.FromSeconds(30),
    FactorySoftTimeout = TimeSpan.FromMilliseconds(100)
  };
  
  public static FusionCacheEntryOptions CategoryOptions = new FusionCacheEntryOptions()
  {
    Duration = TimeSpan.FromMinutes(10),
    IsFailSafeEnabled = true,
    FailSafeMaxDuration = TimeSpan.FromHours(12)
  };
}

[...]

var product = cache.GetOrSet<Product>(
  $"product:{42}",
  _ => GetProductFromDb(42),
  CachingDefaults.ProductOptions // THIS SAME INSTANCE WILL BE PASSED EVERY TIME
);

var category = cache.GetOrSet<Category>(
  $"category:{42}",
  _ => GetCategoryFromDb(42),
  CachingDefaults.CategoryOptions // THIS SAME INSTANCE WILL BE PASSED EVERY TIME
);
```

In this case the `productOptions` or the `categoryOptions` instances will be reused, greatly reducing the number of allocations.
But if we later decides to let the options be adaptive (in any of the proposed ways) and forget to create a new `FusionCacheEntryOptions` object every time, the options instance passed in would be changed, even for the following requests. A practical example is that the ProductOptions is set with a Duration of 1 min, but as soon as an old product will get requested with adaptive caching, the Duration would become 1 hour for all subsequent requests.

To avoid this an idea may be to just duplicate the options object, only in case adaptive caching needs to be applied (which would be possible btw only with Proposal 01, since the adaptive logic is explicitly in a separate param). This would solve the problem, but it would needlessly duplicate those options even in case the instance passed in is already a "throwaway" one. Ideally I should find a way to correctly identify which are reusable and which are not, but I need to experiment a little bit because I can already see some troubling edge cases.

## Thoughts?

Any suggestion or observation would be really, **really** appreciated 🙏.

**Comment by JoeShook** [comment]
In option 1 couldn't the user design the ```GetProductFromDb``` function to do the work of collecting the ```Cache-Control``` header and encapsulating ```product``` or some other wrapper so the lambda would have the data needed for cache duration.  Seems reasonable to give guidance in docs for things like this.  FusionCache is sufficiently complex.  

I have a DNS cache scenario where I need to transform the Cached value.  I could imagine another lambda that allowed me to transform the cached data so I could adjust the TTL (time-to-live) sent back to a client.  But that would add un-needed complexity to FusionCache.  I can easily add business logic after I get the cache to decay the TTL so new client requests are informed of the true TTL, so they can also accurately cache.  What is my point?  I think avoiding more complexity in the FusionCache is be good for these advanced techniques.  

Actually I wonder if a just handling this with a ```Cache.Get``` and ```Cache.Set``` isn't an equivalent solution.  Throwing in the  FusionCacheEntryOptions warning above it would come down to benchmarkdotnet experiments to really decide whether this feature is worth it or if guidance on technique is the way to go.

**Comment by Hawxy** [comment]
I'm a fan of Proposal 2 as it's similar to Microsoft's own caching abstractions and other caching libraries in the ecosystem. It'd allow for much easier "lift n shift" for people coming from other implementations.

Proposal 1 I would be able to get working if I had to. Proposal 3 is a bit of a smell, however I see how it'd meld well with proposal 1.

**Comment by jodydonetti** [comment]
> I'm a fan of Proposal 2 as it's similar to Microsoft's own caching abstractions and other caching libraries in the ecosystem. It'd allow for much easier "lift n shift" for people coming from other implementations.

Good point! I've added a "pro" to Proposal 2.

**Comment by jodydonetti** [comment]
> In option 1 couldn't the user design the `GetProductFromDb` function to do the work of collecting the `Cache-Control` header and encapsulating `product` or some other wrapper so the lambda would have the data needed for cache duration. Seems reasonable to give guidance in docs for things like this. FusionCache is sufficiently complex.

I'm not sure I'm following completely: if I got it right you are saying that the Product class may simply have, for example, a prop with the value of the Cache-Control header, so that it would be available in the second lambda, correct?
If that is the case then yeah, it would absolutely be a solution, but then FusionCache would require you to add a property to your model, which is something someone may feel as a little "invasive".
On the other hand you are using an advanced feature so, I mean, we should put things somewhere right 🤷 ?
Another possible limitation would be with classes you don't control, or sealed, or primitive types. In that case it would not be possible.
But as you said, maybe outlining the limitations in the documentation may be a good enough answer to that.


> What is my point? I think avoiding more complexity in the FusionCache is be good for these advanced techniques.

Totally agree, one of the most important thing in designing something is to strike the right balance between features and complexity, and is not that easy.

**Comment by wileyveteran** [comment]
Ah, this is exactly the kind of thing I was looking for. Want to cache calendar events with the cache duration determined by how far away the event returned is e.g. events happening today might have 1 minute expiry and ones in 6 months 1 day expiry. So did this get abandoned?

**Comment by jodydonetti** [comment]
Hi @wileyveteran , thanks for considering FusionCache!
No, the feature has not been abandoned: the delay has been because, on top of the current situation globally and whatnot, I put myself more into finishing the backplane #11 , which is about to land very soon.
I can schedule this feature soon after that: do you have a specific deadline for having it or you can live without it for a little bit more?

**Comment by wileyveteran** [comment]
We were looking to go to production sometime towards the end of Q1 so we still have a bit of time. Without it I guess we miss out on some of the good features like the cache stampede prevention.

**Comment by jodydonetti** [comment]
Hi all, to move forward with this feature I have to first close the Backplane #11 one, and I'm happy to say that I've finally been able to complete the design and implementation of that.

Please take a look [here](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/releases/tag/v0.1.10-alpha1), try it out and let me know what you think so I can move forward with the final version, and then proceed with the next features (including adaptive caching!).

Thanks everybody 🙏

**Comment by jodydonetti** [comment]
Meanwhile I published an (hopefully) even better [alpha2](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/releases/tag/v0.1.10-alpha2) release.

**Comment by jodydonetti** [comment]
Hi all, just wanted to update you on the next version: I released right now the [BETA2](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/releases/tag/v0.1.10-beta2) for the next big release, which includes among other small things a **big fix for the DI setup part**.

This will probably be the last release before the official one.

**Comment by jodydonetti** [comment]
Hi all, yesterday I released the [BETA 3](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/releases/tag/v0.1.10-beta3).

Unless some big problem comes up, this will be the very last release before the official one, which will be in the next few days 🎉

**Comment by jodydonetti** [comment]
Hi @JoeShook , @wileyveteran and @Hawxy : since I've finally released the backplane with the v0.9 a couple of days ago, I'm now starting to design this very feature.

Just wanted to let you know, and if meanwhile you had some new thoughts about this, please let me know.

**Comment by jodydonetti** [comment]
Hi all, I'd like to quickly update the community on this.

I'm ready to release a new version (maybe a pre-release) with full support for adaptive caching!

I'll give more details asap, but in the meantime I'd just like you to know that I've found an elegant way to go with the (most liked by the community) second option, **with** all the pros, **without** any of the cons, all while keeping the allocations low and optimized 🎉

I think I'll release it with more details in the we, so stay tuned.

**Comment by wileyveteran** [comment]
Looking forward to it, I will try and test it out once you have released it.

**Comment by jodydonetti** [comment]
Hey everybody, I just released [v0.10.0-preview1](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/releases/tag/v0.10.0-preview1) which includes adaptive caching 🎉

If nobody will rise any problem with it I plan on releasing a non pre-release version in a week or so, after I got some time to fully test it in production on a couple of my real-world projects with some high traffic and everything.

The new documentation is available [here](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/blob/feature/adaptive_caching/docs/AdaptiveCaching.md).

For the ones interested, in another comment I'll explain some details on how I have been able to avoid most of the cons of the 2nd proposal being used.

Please let me know what you all think, thanks 🙏


## 📦 It's a pre-release!

Note that the Nuget package is marked as `pre-release` so please be sure to enable the related filter otherwise you would not see them:

![image](https://user-images.githubusercontent.com/1010086/151539799-63f58e84-aa2c-475f-9a9f-05ed5c25904e.png)

**Comment by jodydonetti** [comment]
Ok, it's time to delve into what I've done to prevent most of the cons of proposal 02, for the ones interested in such a thing.

These were the 3 cons emerged during the design phase:

1. **not knowing when options have been changed**: since the adaptive part here would NOT be explicit, we wouldn't know if a factory have in fact changed some options or not, making it harder to save some allocations and avoid unwanted options change
2. **unwanted reusable options changes**: special care must be put into NOT changing a FusionCacheEntryOptions instance when that is not the intention
3. **extra lambda allocation**: when calling the old factory signature (without the context) FusionCache internally calls the new one (with the context) allocating, at the very least, a new lambda for the call indirection

### Points 1 & 2

First thing: I added an `internal` prop to `FusionCacheEntryOptions` so that I can know if an options instance is safe to change via adaptive caching during a factory execution or not. This new property is automatically handled and passed over via duplication, but can be set maually when needed, only internally.

I then put some code in the context's `Options` prop getter to detect when it is accessed, and in that case it makes sure that the options being accessed is safe to be changed, otherwise it automatically duplicates them and uses that instead.

So now we can detect changes and all is safe, avoiding changing a shared `FusionCacheEntryOptions` instance.

It also avoids duplicating options when not necessary, since it only does it when the `Options` prop is accessed, which is fair to think it would be done only when trying to change it.

But only doing this would not be enough, because we would still be duplicating options even in the case of `GetOrSet` calls with an explicit setup action lambda, which already duplicate options starting from the default ones: in these cases it's safe to assume those options are already ok to be changed.

So in everyone of those overloads I also mark them as safe to be changed, resulting in even less allocations.

### Point 3

The third one is more complicated, and as of today there is in fact an extra allocation when calling a `GetOrSet` overload with the old factory signature without the context.

I have a plan for removing this extra allocations, and a bonus point would be to probably be able to remove the extra allocation that is still there when calling a `GetOrSet` overload without a factory and with a default value directly (it is hidden, but it is still there today).

I haven't been able to do it just yet, and since it is an implementation detail that I can change later on I just decided to get out with the new version and see how the community feels about adaptive caching itself, and ideally optimize it later on.

But again, I have a plan for remove these extra allocations, and will work on it asap.

**Comment by JoeShook** [comment]
Hello @jodydonetti .  I updated to  FusionCache 0.10.0-preview1 in my [FusionCache.Metrics](https://github.com/JoeShook/ZiggyCreatures.FusionCache.Metrics) repo.  
With not changes all tests pass.

I started duplicating some of my tests and replacing calls like TryGet with GetOrSet, trying to keep the same behavior and exptecting the same plugin EventHandlers to trigger.  

I have this one test for you to look at where I expected a stale cache hit but it is missing.

````csharp
 // HIT (STALE): +1
                
//cache.TryGet<int>("foo"); 

cache.GetOrSet<int>("foo", (ctx, _) =>
{
    // Duration set by factory return value
    ctx.Options.SetDuration(adaptiveDuration).SetFailSafe(true);

    return 42;
});
````
The code is in the [feature/adaptive_cache branch.](https://github.com/JoeShook/ZiggyCreatures.FusionCache.Metrics/blob/abfcdf36041cbd8454a862c929cb4d9a41d56c75/tests/FusionCache.Plugins.Metrics.EventCounters.Tests/EventCountersTests.cs#L420-L428)

If you swap the GetOrSet back to the orginal TryGet then the expeced stale cache hit is triggered as expected.  

I will continue to experiment with more testing, but this is the first unexpected result I have found so for.

**Comment by jodydonetti** [comment]
Hi @JoeShook that is expected: since you let the cache entry expire, a `TryGet` with fail-safe enabled would activate it, getting you back a stale result (and emitting a stale hit event). But a `GetOrSet` with a factory will see the value as expired, call the factory and set a value, emitting a (I suppose) set event.
Makes sense?

**Comment by JoeShook** [comment]
OK, yes I see what your are saying.  It is indeed a set event and that makes sense in the morning hours :smiley:.  

So, I have another question still in this area.  I modifed the example like the following:

````csharp
// LET IT BECOME STALE
Thread.Sleep(adaptiveDuration);

// HIT (STALE): +1
cache.TryGet<int>("foo");
// HIT (STALE): +1
cache.TryGet<int>("foo");
````
I expected two stale hit events.  But I get one stale hit event and one hit event.  Is that by design?  Locically I expected to continue to get a stale cache values until I set the expired cache via one of the set methods. 

It seems that the first stale result promotes the stale cache to current non-stale result.  At least the events say that is the case.

**Comment by JoeShook** [comment]
Oh, I see now.  It is by design.  The FailSafeThrottleDuration is a play here to avoid excessive load on the distributed cache.  I am a little stale 😐 on my FusionCache knowledge.

**Comment by jodydonetti** [comment]
HI @JoeShook yes, you got it right!
But now that you mention it, it may make sense to reason differently: shouldn't a cache hit of an entry that came from a fail-safe activation co

**Comment by jodydonetti** [comment]
**PS:** I think I'll open 2 new issues related to a behaviour change, one for the point you just made (stale cache hit for entries re-saved in the past via a fail-safe activation) and another one to maybe not activate fail-safe without a factory (see [here](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/issues/45#issuecomment-1102290558)). Initial thoughts?

**Comment by Hawxy** [comment]
Swapped LazyCache for the FusionCache preview in one of my libraries that needs adaptive caching and the tests are green. Seems good to me. 👍

**Comment by JoeShook** [comment]
> HI @JoeShook yes, you got it right! But now that you mention it, it may make sense to reason differently: shouldn't a cache hit of an entry that came from a fail-safe activation be considered stale? I mean, not just when fail-safe is being activated (because for that we alredy have a specific event), but every time a cache entry that is in fact stale is returned?
> 
> You may have a point there, you know? Let me think about it.

Yes, I did pause after my comment.  I was thinking of responding with the same observation you mentioned.  I think once the cache is stale it should event a Hit with the IsStale argument property until it is refreshed from a successful factory result.

**Comment by jodydonetti** [comment]
I just released [v0.10.0](https://github.com/jodydonetti/ZiggyCreatures.FusionCache/releases/tag/v0.10.0) with adaptive caching 🎉

It also includes the change for cache hit events as suggested by @JoeShook and more.

Please try it out if you can and let me know!

**Comment by JoeShook** [comment]
It works great.  I released [v1.0.4](https://github.com/JoeShook/ZiggyCreatures.FusionCache.Metrics/releases/tag/v1.0.4) 

FYI:  Working on the OpenTelemetry plugin now.  The plugin is simple but the examples will take some time.  Working in the [otelWithExample](https://github.com/JoeShook/ZiggyCreatures.FusionCache.Metrics/tree/otelWithExample) branch.
